GH-32747: [C++] Substrait To Arrow Emit feature testing

vibhatha opened the pull request:
Adding compound tests for Substrait Emit. This is an extension to previous tests which only included emit testing only with a particular operator + read.

Also the util functions used in tests are moved to `test_util.cc` where future tests including compound relations can re-use this logic.
* Closes: #32747

westonpace commented on cpp/src/arrow/engine/substrait/serde_test.cc:
```diff
@@ -3187,5 +3043,637 @@ TEST(Substrait, IsthmusPlan) {
+}
+
+TEST(Substrait, CompoundEmitWithFilter) {
+#ifdef _WIN32
+  GTEST_SKIP() << "ARROW-16392: Substrait File URI not supported for Windows";
+#endif
```
Same comment as above.

westonpace commented on cpp/src/arrow/engine/substrait/test_util.cc:
```diff
@@ -0,0 +1,117 @@
+  ASSERT_OK_AND_ASSIGN(auto sink_decls, DeserializePlans(
+                                            *buf, [] { return kNullConsumer; },
+                                            ext_id_reg, &ext_set, conversion_options));
+  auto other_declrs = std::get_if<compute::Declaration>(&sink_decls[0].inputs[0]);
```
```suggestion
  auto other_declrs = std::get_if<compute::Declaration>(&sink_decls[0].inputs[0]);
  ASSERT_NE(other_declrs, nullptr);
```

vibhatha replied:
How about 
`ASSERT_TRUE(other_declrs.IsValid());`

westonpace reviewed (changes requested):
Just minor nits outside of my usual concern about having so many hard-coded substrait JSON plans.  One thing I think we should look into is whether or not we still need to skip due to ARROW-16392 (which is resolved).  If it isn't working then we should either fix it here or update our JIRA #'s.

vibhatha commented on cpp/src/arrow/engine/substrait/serde_test.cc:
```diff
@@ -3187,5 +3043,637 @@ TEST(Substrait, IsthmusPlan) {
 }
 
+TEST(Substrait, CompoundEmitFilterless) {
+#ifdef _WIN32
+  GTEST_SKIP() << "ARROW-16392: Substrait File URI not supported for Windows";
+#endif
```
@westonpace sorry I made the PR https://github.com/apache/arrow/pull/14425 and the JIRA is here, https://issues.apache.org/jira/browse/ARROW-18051

vibhatha commented on cpp/src/arrow/engine/substrait/serde_test.cc:
```diff
@@ -3187,5 +3043,637 @@ TEST(Substrait, IsthmusPlan) {
+            output_table = right_table;
+          }
+        }
+        std::shared_ptr<compute::ExecNodeOptions> options =
```
Not sure if we can do that here since the lambda returns a `Result<Declaration>`, but maybe

```c++
if(!output_table) {
     return Status::Invalid("NamedTableProvider couldn't set the table");
}
```

vibhatha commented:
cc @westonpace addressed the reviews

westonpace commented on cpp/src/arrow/engine/substrait/serde_test.cc:
```diff
@@ -3692,5 +3536,641 @@ TEST(Substrait, ReadRelWithGlobFiles) {
+  ASSERT_OK_AND_ASSIGN(auto buf, internal::SubstraitFromJSON("Plan", substrait_json));
+  auto output_schema = schema({
+      field("A", int32()),
+      field("E", int32()),
```
```suggestion
      field("C", int32()),
```

westonpace commented on cpp/src/arrow/engine/substrait/serde_test.cc:
```diff
@@ -3692,5 +3536,641 @@ TEST(Substrait, ReadRelWithGlobFiles) {
+      field("E", int32()),
+      field("B+C", int32()),
+      field("X", int32()),
+      field("Z", int32()),
```
```suggestion
      field("W", int32()),
```

westonpace commented on cpp/src/arrow/engine/substrait/test_util.h:
```diff
@@ -0,0 +1,118 @@
+    field("ts_ns", timestamp(TimeUnit::NANO)),
+});
+
+ARROW_ENGINE_EXPORT std::shared_ptr<DataType> StripFieldNames(
+    std::shared_ptr<DataType> type);
+
+inline compute::Expression UseBoringRefs(const compute::Expression& expr) {
+  if (expr.literal()) return expr;
+
+  if (auto ref = expr.field_ref()) {
+    return compute::field_ref(*ref->FindOne(*kBoringSchema));
+  }
+
+  auto modified_call = *CallNotNull(expr);
+  for (auto& arg : modified_call.arguments) {
+    arg = UseBoringRefs(arg);
+  }
+  return compute::Expression{std::move(modified_call)};
+}
```
Why did these two methods have to move to a header file?

vibhatha replied:
I assumed util. This is the move which causes the windows build failure.

vibhatha replied:
I undo the move.

westonpace reviewed (commented):
Can you rebase this?

westonpace commented on cpp/src/arrow/engine/substrait/test_util.cc:
```diff
@@ -0,0 +1,129 @@
+void WriteIpcData(const std::string& path,
+                  const std::shared_ptr<fs::FileSystem> file_system,
+                  const std::shared_ptr<Table> input) {
+  EXPECT_OK_AND_ASSIGN(auto mmap, file_system->OpenOutputStream(path));
+  ASSERT_OK_AND_ASSIGN(
+      auto file_writer,
+      MakeFileWriter(mmap, input->schema(), ipc::IpcWriteOptions::Defaults()));
+  TableBatchReader reader(input);
+  std::shared_ptr<RecordBatch> batch;
+  while (true) {
+    ASSERT_OK(reader.ReadNext(&batch));
+    if (batch == nullptr) {
+      break;
+    }
+    ASSERT_OK(file_writer->WriteRecordBatch(*batch));
+  }
```
```suggestion
  EXPECT_OK_AND_ASSIGN(auto out_stream, file_system->OpenOutputStream(path));
  ASSERT_OK_AND_ASSIGN(
      auto file_writer,
      MakeFileWriter(out_stream, input->schema(), ipc::IpcWriteOptions::Defaults()));
  EXPECT_OK(file_writer->WriteTable(*input));
```
There is a `WriteTable` method that will simplify this.

vibhatha replied:
You're correct. I just rebased this today. I will add your suggestion.

vibhatha replied:
`ASSERT_OK(file_writer->Close());` 
Need this too, I think.

westonpace commented on cpp/src/arrow/engine/substrait/test_util.cc:
```diff
@@ -0,0 +1,129 @@
+  ASSERT_OK(file_writer->Close());
+}
+
+Result<std::shared_ptr<Table>> GetTableFromPlan(
```
This method is no longer needed now that we have `DeclarationToTable`.

vibhatha replied:
Ah, let me take a look.

vibhatha replied:
That's correct. I am removing this.

westonpace commented on cpp/src/arrow/engine/substrait/test_util.cc:
```diff
@@ -0,0 +1,129 @@
+  std::shared_ptr<ExtensionIdRegistry> sp_ext_id_reg = MakeExtensionIdRegistry();
+  ExtensionIdRegistry* ext_id_reg = sp_ext_id_reg.get();
+  ExtensionSet ext_set(ext_id_reg);
+  ASSERT_OK_AND_ASSIGN(auto sink_decls, DeserializePlans(
+                                            *buf, [] { return kNullConsumer; },
+                                            ext_id_reg, &ext_set, conversion_options));
+  auto& other_declrs = std::get<compute::Declaration>(sink_decls[0].inputs[0]);
```
There is now a `DeserializePlan` method that returns a single declaration.

vibhatha replied:
This one causes a segmentation fault in one of the test cases which is already merged (`Substrait.NestedProjectWithMultiFieldExpressions`), I will probably investigate this in a separate ticket.

vibhatha replied:
https://github.com/apache/arrow/issues/33844

vibhatha replied:
I have assigned this one to myself.

westonpace commented on cpp/src/arrow/engine/substrait/test_util.h:
```diff
@@ -0,0 +1,59 @@
+
+const auto kNullConsumer = std::make_shared<compute::NullSinkNodeConsumer>();
+
+ARROW_ENGINE_EXPORT std::shared_ptr<DataType> StripFieldNames(
+    std::shared_ptr<DataType> type);
```
Is this used anymore?

vibhatha replied:
You're right, this can be removed.

vibhatha replied:
You're correct, we don't need this now.

westonpace reviewed (changes requested):
Sorry for the delay in getting back to this.  Some of these utilities can be simplified now that we've made things simpler elsewhere.

vibhatha commented:
@westonpace I updated the PR. Some windows CIs are failing. Could it be because I removed the windows skip flag on `ReadRelWithGlobFiles` test case?

westonpace commented on cpp/src/arrow/engine/substrait/serde_test.cc:
```diff
@@ -88,26 +88,53 @@ using internal::checked_cast;
+  }
+  ASSERT_OK_AND_ASSIGN(output_table, output_table->CombineChunks());
+  ASSERT_OK_AND_ASSIGN(auto merged_expected, expected_table->CombineChunks());
+  std::cout << merged_expected->ToString() << std::endl;
+  std::cout << "----------------------" << std::endl;
+  std::cout << output_table->ToString() << std::endl;
```
We should probably remove these prints

westonpace commented:
> @westonpace I updated the PR. Some windows CIs are failing. Could it be because I removed the windows skip flag on ReadRelWithGlobFiles test case?

Yes, I believe we still need to skip that test on Windows.  Could you open up a new issue to investigate?

vibhatha commented:
> > @westonpace I updated the PR. Some windows CIs are failing. Could it be because I removed the windows skip flag on ReadRelWithGlobFiles test case?
> 
> Yes, I believe we still need to skip that test on Windows. Could you open up a new issue to investigate?

@westonpace Issue created here: https://github.com/apache/arrow/issues/33861

vibhatha commented:
@westonpace seems like the CIs are passing now. But one CI's failing and it could be unrelated.

westonpace commented:
> @westonpace seems like the CIs are passing now. But one CI's failing and it could be unrelated.

I agree those failures are unrelated.
